driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -89,51 +83,71 @@ impl StableXOrderBookReading for UpdatingOrderbook {
 /// Returns Ok when exit_indicator is dropped.
 /// Returns Err if the stream ends.
 async fn update_with_events_forever(
+    web3: Web3,
+    contract: Arc<dyn StableXContract>,
     orderbook: Arc<Mutex<Orderbook>>,
     orderbook_ready: Arc<AtomicBool>,
     mut block_timestamp_reader: impl BlockTimestampReading,
     exit_indicator: oneshot::Receiver<()>,
-    past_events: impl Future<Output = Result<Vec<Event<batch_exchange::Event>>, ExecutionError>>,
-    stream: impl Stream<Item = Result<Event<batch_exchange::Event>, ExecutionError>>,
 ) -> Result<()> {
-    // `select!` requires the futures to be fused...
+    const POLL_INTERVALL: Duration = Duration::from_secs(30);
+    const BLOCK_RANGE: u64 = 25;
+
+    log::info!("Starting event based orderbook updating.");
+
+    // `select!` requires the futures to be fused and pinned.
     let exit_indicator = exit_indicator.fuse();
-    let past_events = past_events.fuse();
-    let stream = stream.fuse();
-    // ...and pinned.
     pin_mut!(exit_indicator);
-    pin_mut!(past_events);
-    pin_mut!(stream);
 
-    log::info!("Starting event based orderbook updating.");
+    select_biased! {
+        _ = exit_indicator => return Ok(()),
+        events = contract.events(BlockNumber::Earliest, BlockNumber::Latest).fuse() => {
+            handle_events(&orderbook, &mut block_timestamp_reader, events?, 0).await?;
+            orderbook_ready.store(true, Ordering::SeqCst);
+        },
+    }
 
     loop {
-        // We select over everything together instead of for example the past events first then the
-        // stream to ensure that the stream gets polled at least once which it needs in order to
-        // create the corresponding filter on the node.
+        // TODO: This is not optimal because it means we might sleep for some time when exit
+        // indicator triggered. We should create a sleep future that expires after poll interval and
+        // join both together.
+        std::thread::sleep(POLL_INTERVALL);
+        let current_block = web3.eth().block_number().compat().await?;
+        let from_block = current_block.saturating_sub(From::from(BLOCK_RANGE));
```

Comment:
I'm wondering what happens if the application of all state takes longer than 25 blocks (which on my machine is definitely the case). Would it make sense to have this be the min of current - BLOCK_RANGE and the maximum existing block number?

Reply:
Added a fix for this.